### PR TITLE
[Sunrise] Add cuda-graph, int8 and more model support on Sunrise

### PR DESCRIPTION
**Pull request by cyx11111**
### PR Category
Vendor

### PR Type
New Features

### Description
This PR introduce support of cuda-graph and int8 with vllm 0.20.2 on Sunrise.
The support list of model is also increased. By now the supported mode is included but not limited on:
Qwen3/3.5/3.6
MiniCPM-o 4.5

### Testing
To test models on this PR with latest sunrise software, you should:

1. use python=3.10
2. upgrade FlagCX to 0.13.0
3. use the latest FlagGems
4. add the following ENV:
export TA_CLOSE_GRAPH_MEMORY=1
export LIBRARY_PATH=/usr/local/tangrt/targets/linux-x86_64/lib:/usr/local/tangrt/lib/linux-x86_64:${LIBRARY_PATH}
export LD_LIBRARY_PATH=/usr/local/tangrt/targets/linux-x86_64/lib:/usr/local/tangrt/lib/linux-x86_64:${LD_LIBRARY_PATH}

1. If you start a model with no VL support, here is a example command:
vllm serve /path/to/Qwen3-32B --tensor-parallel-size 2 --max-num-batched-tokens 16384 --max-num-seqs 16 --gpu-memory-utilization 0.8 --port 7529 --compilation-config '{"mode": 0, "cudagraph_mode": "FULL_DECODE_ONLY"}' 

2. If you start a model with VL (such as Qwen3.5/3.6), here is a example command:
vllm serve /path/to/Qwen3.5-35B-A3B --tensor-parallel-size 2 --max-num-batched-tokens 16384 --max-num-seqs 16 --gpu-memory-utilization 0.8 --port 7529 --skip-mm-profiling --compilation-config '{"mode": 0, "cudagraph_mode": "FULL_DECODE_ONLY"}'

3. If you want to try INT8 path, you need first use FlagOS-Compressor to convert w8a8 model: 
flagos-compressor quantize \
  --input  /path/to/Qwen3.5-35B-A3B \
  --output /path/to/Qwen3.5-35B-A3B-w8a8-flagos \
  --select moe \
  --bits 8 \
  --activation-bits 8 \
  --strategy channel \
  --method mse \
  --n-candidates 200 \
  --chunk-size 1024 \
  --backend cpu
  And then use vllm serve to start model

### Checklist
- [ ] I have run the existing tests and they pass
- [ ] I have added tests for my changes (if applicable)
- [ ] I have updated the documentation (if applicable)

### CODE REVIEW

File: vllm_fl/dispatch/backends/vendor/sunrise/impl/mla.py [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+
+        # flash_mla returns [B, s_q, H, dv]; common layer wants [B, H, dv]
+        # (s_q==1 for pure decode). Pass vLLM softmax scale when supported.
+        # Current Sunrise FlagGems ``flash_mla`` hardcodes 1/sqrt(d); optional
+        # ``sm_scale`` is accepted when newer FlagGems exposes it.
+        flash_mla_kwargs = {}
+        try:
+            import inspect
+
+            if "sm_scale" in inspect.signature(flash_mla).parameters:
```
Comment (cyber-pioneer) [inline vllm_fl/dispatch/backends/vendor/sunrise/impl/mla.py]: 这里可能导致 MLA decode 结果不正确。当前只有在 `flash_mla` 暴露 `sm_scale` 参数时才传入 `self.scale`，否则会静默使用 Sunrise FlagGems kernel 内部硬编码的 `1 / sqrt(d)`。对于 MLA，`d` 是吸收后的维度（通常为 `kv_lora_rank + qk_rope_head_dim`），而 `self.scale` 根据模型原始 QK head dimension 计算，两者通常不同，因此 fallback 会改变 attention logits。建议让 kernel 支持并始终传入 `self.scale`；如果当前版本不支持，应明确报错，而不是使用不同的计算语义。

File: vllm_fl/dispatch/backends/vendor/sunrise/impl/mla.py [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+
+        if self.kv_cache_dtype.startswith("fp8"):
+            raise NotImplementedError("FP8 Sunrise MLA not yet supported")
+
+        if isinstance(q, tuple):
+            q = torch.cat(q, dim=-1)
+
+        assert isinstance(q, torch.Tensor)
+        # Common path: [B, H, D]. flash_mla wants [B, s_q, H, D].
+        if q.ndim == 3:
```
Comment (cyber-pioneer) [inline vllm_fl/dispatch/backends/vendor/sunrise/impl/mla.py]: 三维 query 不一定表示每个请求只有一个 token。speculative/multi-token decode 中，vLLM 会传入展平后的 `[num_decodes * query_len, num_heads, head_dim]`。这里直接把第一维作为 batch 并设置 `s_q = 1`，当 `query_len > 1` 时，`b` 会与 `block_table`、`seq_lens` 的请求维度不一致，可能造成 KV cache 映射错误或 kernel 失败。建议使用 `attn_metadata.num_decodes` 将 query reshape 为 `[num_decodes, -1, H, D]`；如果 Sunrise `flash_mla` 暂不支持 `s_q > 1`，应明确拒绝 speculative decode，并增加 `query_len >= 2` 的测试。

File: vllm_fl/dispatch/backends/vendor/sunrise/impl/fla/solve_tril.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+_PTPU_SOLVE_TRIL_DISABLED = False
+
+
+def _fla_fallback(
+    A: torch.Tensor,
+    cu_seqlens: Optional[torch.Tensor],
+    chunk_indices: Optional[torch.Tensor],
+    output_dtype: torch.dtype,
+) -> torch.Tensor:
+    from vllm.model_executor.layers.fla.ops.solve_tril import (
```
Comment (cyber-pioneer) [inline vllm_fl/dispatch/backends/vendor/sunrise/impl/fla/solve_tril.py]: 这个 fallback 无法调用原始 FLA 实现。`patch_fla_ops.py` 已将模块中的 `solve_tril` 替换为 PTPU wrapper，因此这里再次从同一模块导入时，得到的仍然是当前 wrapper；fallback 会再次进入自身并最终触发 `RecursionError`。同样的问题也存在于 `chunk_fwd_o`、`chunk_h`、`chunk_scaled_dot_kkt` 和 `cumsum`。建议在 monkey patch 前保存所有原始函数引用，并在 fallback 中直接调用这些引用，就像当前对 fused sigmoid 和 fused recurrent 的处理一样；同时增加主动触发各 fallback 路径的测试。

**Review by cyber-pioneer — CHANGES_REQUESTED** [review]
本 PR 目前存在 3 个需要在合并前修复的正确性问题：MLA decode 可能使用错误的 softmax scale；query reshape 不符合 vLLM 的 speculative/multi-token decode 约定；多个 FLA fallback 会递归调用被 patch 后的 wrapper。

File: vllm_fl/dispatch/backends/vendor/sunrise/impl/mla.py (hunk 2 above)
Comment (cyx11111) [reply vllm_fl/dispatch/backends/vendor/sunrise/impl/mla.py]: fix in 010fc7a

File: vllm_fl/dispatch/backends/vendor/sunrise/impl/mla.py (hunk 1 above)
Comment (cyx11111) [reply vllm_fl/dispatch/backends/vendor/sunrise/impl/mla.py]: fix in 010fc7a

File: vllm_fl/dispatch/backends/vendor/sunrise/impl/fla/solve_tril.py (hunk 3 above)
Comment (cyx11111) [reply vllm_fl/dispatch/backends/vendor/sunrise/impl/fla/solve_tril.py]: fix in f25b510 and ba8ccfd

**Review by rdzhu225 — APPROVED** [review]

**Review by physics31415926 — APPROVED** [review]

File: vllm_fl/platform.py [hunk 4]
```diff
@@ -464,6 +468,8 @@ def pre_register_and_update(cls, parser=None) -> None:
             import vllm_fl.dispatch.backends.vendor.ascend
         elif cls.device_name == "gcu":
             import vllm_fl.dispatch.backends.vendor.gcu  # noqa: F401
+        elif cls.device_type == "ptpu":
+            import vllm_fl.dispatch.backends.vendor.sunrise  # noqa: F401
```
Comment (ceci3) [inline vllm_fl/platform.py:473]: why need import vllm_fl.dispatch.backends.vendor.sunrise   twice?

**Review by cyber-pioneer — APPROVED** [review]

**Comment by physics31415926** [comment]
@cyx11111 你好，我们在排查 release/0.2 → main 的同步需求。release/0.2 对应 vLLM 0.20.2，当前 main 对应 vLLM 0.24.0。

本 PR 于 8 月 31 日合入，包含 Sunrise CUDA Graph、INT8 和更多模型支持；现有面向 0.24 的 #391 主要处理 attention backend 注册和 memory_stats 兼容，未覆盖这批扩展。另有 #475 处理 FlagCX wrapper API 的单点修复。

麻烦评估本 PR 中哪些能力需要同步到 main，以及哪些 FLA/GDN/MoE、图捕获补丁需要按 0.24 接口重新适配。改动较大，建议按功能拆分，并结合 Sunrise 硬件验证决定迁移范围，避免整包直接 cherry-pick。谢谢！
